openlibrary/core/bookshelves.py

```diff
@@ -202,6 +202,97 @@ def count_total_books_logged_by_user_per_shelf(
+    # doc["editions"]["docs"]
+    def link_editions_to_works(solr_docs):
+        linked_docs: list[web.storage] = []
+        docs_dict = {}
+        # adds works to linked_docs, recording their edition key and index in docs_dict if present.
+        for doc in solr_docs:
+            full_key = doc.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "works":
+                linked_docs.append(doc)
+                if doc.get("logged_edition"):
+                    docs_dict.update(
+                        {doc["logged_edition"].split("/")[2]: (len(linked_docs) - 1)}
+                    )
+        # Attaches editions to the works, in second loop-- in case of misperformed order.
+        for edition in solr_docs:
+            full_key = edition.get("key") or "/dummy/value"
+            _, key, item_id = full_key.split("/")
+            if key == "books":
+                index = docs_dict.get(item_id, -1)
+                if index >= 0:
+                    linked_docs[index].editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning("Error: No work found for edition %s" % full_key)
+        logger.info("LOGGERDICT IS %s" % docs_dict)
+        return linked_docs
+
+    @classmethod
+    def add_storage_items_for_redirects(
+        cls,
+        reading_log_work_keys,
+        solr_docs: list[web.Storage],
+        query_params,
+        limit,
+    ) -> list[web.storage]:
+        """
+        Use reading_log_work_keys to fill in missing redirected items in the
+        the solr_docs query results.
+
+        Solr won't return matches for work keys that have been redirected. Because
+        we use Solr to build the lists of storage items that ultimately gets passed
+        to the templates, redirected items returned from the reading log DB will
+        'disappear' when not returned by Solr. This remedies that by filling in
+        dummy works, albeit with the correct work_id.
+        """
+
+        from openlibrary.plugins.worksearch.code import run_solr_query
+        from openlibrary.plugins.worksearch.schemes.works import WorkSearchScheme
+
+        work_edition_dict = {
+            work_key: edition_key.split("/")[2] if edition_key else None
+            for (work_key, edition_key) in reading_log_work_keys
+        }
+        reading_log_work_keys = [keys[0] for keys in reading_log_work_keys if keys[0]]
+        for idx, work_key in enumerate(reading_log_work_keys):
+            corresponding_solr_doc = next(
+                (doc for doc in solr_docs if doc.key == work_key), None
+            )
+
+            if not corresponding_solr_doc:
+                logger.info("failed key: %s" % work_key)
+                edition_key = work_edition_dict.get(work_key)
+                solr_resp = (
+                    run_solr_query(
```

Comment:
Having a solr query for every item here is less ideal performance wise since it means the requests the page makes is O(n) relative to the number redirects!

Maybe something like this:

```py
solr_docs
reading_log_work_keys

key_to_solr_doc = {doc[“key”]: doc for doc in solr_docs}

fetched_keys = set(doc[‘key’] for doc in solr_doc)
missing_keys = reading_log_work_keys - fetched_keys

new_docs = solr_get_many(missing_keys, ...)
Key_to_new_doc = {doc[“key”]: doc for doc in new_docs}

return [
   solr_docs.get(key) or new_docs.get(key)
    For key in reading_log_work_keys
]
```

Reply:
There's a lot of optimization that I'm still learning about; I was aware of O(n) notation, for example, but I wasn't sure when we particularly cared about it. The number of requests made is a perfect scenario for this, and I'm glad that the current version now packages all these requests in a batch.